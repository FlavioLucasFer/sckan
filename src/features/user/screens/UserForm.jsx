import React, { useState, useEffect } from 'react';
import { useLocation } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';


import FormScreen from 'core/components/FormScreen';
import useFormInput from 'core/hooks/useFormInput';
import TextInput from 'core/components/TextInput';
import Select from 'core/components/Select';
import { USER_ROUTE } from 'core/utils/routes';

import userService from 'features/user/domain/service';
import userModel from 'features/user/domain/model';

import { setUsers } from 'redux/reducers/users';

function UserForm() {
	const dispatch = useDispatch();
	const location = useLocation();
	
	const [user, setUser] = useState({});
	const name = useFormInput();
	const login = useFormInput();
	const password = useFormInput();
	const cpf = useFormInput();
	const companyPosition = useFormInput();
	const company = useFormInput();
	const [passwordIsVisible, setPasswordIsVisible] = useState(false);
	
	const companies = useSelector(state => state.companies.value);

	useEffect(() => {
		if (location.state) {
			function fakeTarget(value) {
				return {
					target: {
						value,
					},
				};
			}
			
			const user = location.state.data;
			setUser(user);
			
			name.onChange(fakeTarget(user.name));
			login.onChange(fakeTarget(user.login));
			cpf.onChange(fakeTarget(user.cpf));
			companyPosition.onChange(fakeTarget(user.companyPosition));
			company.onChange(fakeTarget(user.company));
		}
	}, []);

	const companyPositions = [
		{ title: 'Scrum Master', value: 'SM' },
		{ title: 'Product Owner', value: 'PO' },
		{ title: 'Developer', value: 'DV' },
		{ title: 'Admin', value: 'AD' },
	];

	function handlePasswordVisilityIconClick() {
		setPasswordIsVisible(!passwordIsVisible);
	}

	async function handleSave() {
		const User = userModel(
			user.id,
			name.value,
			login.value,
			password.value,
			cpf.value,
			companyPosition.value,
			company.value,
		);

		try {
			await userService.save(User);
			await refreshUsers();
		} catch (err) {
			console.log('err:', err);
		}
	}

	async function refreshUsers() {
		const users = await userService.find();
		dispatch(setUsers(users));
	}

	return (
		<FormScreen title="User"
			id={user.id}
			backRoute={USER_ROUTE}
			onSave={handleSave}>
			<div className="row">
				<TextInput {...name}
					title="Name"
					className="col s12 m6 l8"
					type="text"
					icon="account_box"
					validate
					mandatory
					required />

				<TextInput {...cpf}
					title="CPF"
					className="col s12 m6 l4"
					type="text"
					icon="assignment_ind"
					mask="cpf"
					validateBy="cpf"
					validate
					mandatory
					required />

				<TextInput {...login}
					title="Login"
					className="col s12 m6"
					type="text"
					icon="login"
					validate
					mandatory
					required />

				<TextInput {...password}
					title="Password"
					className="col s12 m6"
					type={passwordIsVisible ? 'text' : 'password'}
					icon={passwordIsVisible ? 'visibility_off' : 'visibility'}
					validate={!user}
					mandatory={!user}
					required={!user}
					iconClick={handlePasswordVisilityIconClick} />
			</div>

			<div className="row">
				<Select {...companyPosition}
					title="Company position"
					className="col s12 m6 l4"
					values={companyPositions}
					defaultOption="Select a company position"
					icon="badge" />

				<Select {...company}
					title="Company"
					className="col s12 m6 l4"
					values={companies}
					valueAs="id"
					titleAs="companyName"
					defaultOption="Select a company"
					icon="business" />
			</div>
		</FormScreen>
	);
}

export default UserForm;
