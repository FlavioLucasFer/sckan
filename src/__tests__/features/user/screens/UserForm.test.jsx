import React from 'react';
import { shallow } from 'enzyme';

import UserForm from 'features/user/screens/UserForm';

describe('Testing UserForm form component', () => {
	it('should render correctly', () => {
		const wrapper = shallow(
			<UserForm />
		);

		expect(wrapper.debug()).toMatchSnapshot();

		wrapper.setProps({
			id: '98890890',
		});

		expect(wrapper.debug()).toMatchSnapshot();
	});

	it('should exchange icon and input type when click on password input icon', () => {
		const wrapper = shallow(
			<UserForm />
		);

		expect(
			wrapper.find({ title: 'Password' })
				.props()
				.type
		).toBe('password');

		expect(
			wrapper.find({ title: 'Password' })
				.props()
				.icon
		).toBe('visibility');


		wrapper.find({ title: 'Password' })
			.dive()
			.find('Icon')
			.dive()
			.simulate('click');

		expect(
			wrapper.find({ title: 'Password' })
				.props()
				.type
		).toBe('text');

		expect(
			wrapper.find({ title: 'Password' })
				.props()
				.icon
		).toBe('visibility_off');

		wrapper.find({ title: 'Password' })
			.dive()
			.find('Icon')
			.dive()
			.simulate('click');

		expect(
			wrapper.find({ title: 'Password' })
				.props()
				.type
		).toBe('password');

		expect(
			wrapper.find({ title: 'Password' })
				.props()
				.icon
		).toBe('visibility');
	});
});