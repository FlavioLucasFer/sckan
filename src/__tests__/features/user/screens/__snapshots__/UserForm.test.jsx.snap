// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing UserForm form component should render correctly 1`] = `
"<FormScreen title=\\"User\\" id={[undefined]}>
  <div className=\\"row\\">
    <TextInput value=\\"\\" onChange={[Function: handleChange]} title=\\"Name\\" className=\\"col s12 m6 l8\\" type=\\"text\\" icon=\\"account_box\\" validate={true} mandatory={true} required={true} />
    <TextInput value=\\"\\" onChange={[Function: handleChange]} title=\\"CPF\\" className=\\"col s12 m6 l4\\" type=\\"text\\" icon=\\"assignment_ind\\" mask=\\"cpf\\" validateBy=\\"cpf\\" validate={true} mandatory={true} required={true} />
    <TextInput value=\\"\\" onChange={[Function: handleChange]} title=\\"Login\\" className=\\"col s12 m6\\" type=\\"text\\" icon=\\"login\\" validate={true} mandatory={true} required={true} />
    <TextInput value=\\"\\" onChange={[Function: handleChange]} title=\\"Password\\" className=\\"col s12 m6\\" type=\\"password\\" icon=\\"visibility\\" validate={true} mandatory={true} required={true} iconClick={[Function: handlePasswordVisilityIconClick]} />
  </div>
  <div className=\\"row\\">
    <Select value=\\"\\" onChange={[Function: handleChange]} title=\\"Company position\\" className=\\"col s12 m6 l4\\" values={{...}} defaultOption=\\"Select a company position\\" icon=\\"badge\\" />
    <Select value=\\"\\" onChange={[Function: handleChange]} title=\\"Company\\" className=\\"col s12 m6 l4\\" values={{...}} defaultOption=\\"Select a company\\" icon=\\"business\\" />
  </div>
</FormScreen>"
`;

exports[`Testing UserForm form component should render correctly 2`] = `
"<FormScreen title=\\"User\\" id=\\"98890890\\">
  <div className=\\"row\\">
    <TextInput value=\\"\\" onChange={[Function: handleChange]} title=\\"Name\\" className=\\"col s12 m6 l8\\" type=\\"text\\" icon=\\"account_box\\" validate={true} mandatory={true} required={true} />
    <TextInput value=\\"\\" onChange={[Function: handleChange]} title=\\"CPF\\" className=\\"col s12 m6 l4\\" type=\\"text\\" icon=\\"assignment_ind\\" mask=\\"cpf\\" validateBy=\\"cpf\\" validate={true} mandatory={true} required={true} />
    <TextInput value=\\"\\" onChange={[Function: handleChange]} title=\\"Login\\" className=\\"col s12 m6\\" type=\\"text\\" icon=\\"login\\" validate={true} mandatory={true} required={true} />
    <TextInput value=\\"\\" onChange={[Function: handleChange]} title=\\"Password\\" className=\\"col s12 m6\\" type=\\"password\\" icon=\\"visibility\\" validate={true} mandatory={true} required={true} iconClick={[Function: handlePasswordVisilityIconClick]} />
  </div>
  <div className=\\"row\\">
    <Select value=\\"\\" onChange={[Function: handleChange]} title=\\"Company position\\" className=\\"col s12 m6 l4\\" values={{...}} defaultOption=\\"Select a company position\\" icon=\\"badge\\" />
    <Select value=\\"\\" onChange={[Function: handleChange]} title=\\"Company\\" className=\\"col s12 m6 l4\\" values={{...}} defaultOption=\\"Select a company\\" icon=\\"business\\" />
  </div>
</FormScreen>"
`;
